src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -527,6 +537,7 @@ public void setCharAt(int index, char ch) {
                 inflate();
             }
             StringUTF16.putCharSB(value, index, ch);
+            maybeLatin1 = true;
```

Comment:
Only possible if `StringLatin1.canEncode(ch) == true` (and strictly speaking if the char at `index` is not latin1-encodable). Considering the overheads of adding a test it might be preferable to be conservative and just always mark it as a "maybe" like you do, though.